test/test_sparse.py

```diff
@@ -1147,6 +1125,45 @@ def do_test(t):
 
         do_test(self.SparseTensor())
 
+    def _test_resize_shape(self, x_i, x_v, x_size, y_i, y_v, y_size):
+        x = torch.sparse_coo_tensor(torch.zeros(x_i), torch.ones(x_v), torch.Size(x_size))
+        y = torch.sparse_coo_tensor(torch.zeros(y_i), torch.ones(y_v), torch.Size(y_size))
+        x.resize_as_(y)
+        self.assertEqual(x.shape, y.shape)
+        self.assertEqual(x._sparseDims(), y._sparseDims())
+        self.assertEqual(x._denseDims(), y._denseDims())
```

Comment:
is the expectation that these resize_as_ calls behave equivalently as for dense tensors?  If so, you should check this.  FYI when I check this I get that case 2, 3, and 5 fail (caveat: I may not be checking this correctly, I just hacked something up).  It's probably also better to do randn rather than ones for the values, since using ones everywhere can cover up issues.

Reply:
There seems to be two issues if we do the value check against a similarly resized dense tensors for equivalence:

1. When we expand the size of the `values` tensor, since the newly allocated region is uninitialized (and not always set to zero), we can't guarantee that they have the same value in both the sparse tensor and the dense tensor we are comparing against.

2. The memory layout of a sparse tensor's `values` and the equivalent dense tensor is different, and the same resize operation on them can give different results. For example:

```
>>> x_i, x_v, x_size = [1, 1], [1, 2, 3], [2, 2, 3]
>>> y_i, y_v, y_size = [1, 1], [1, 2], [2, 2]
>>> x = torch.sparse_coo_tensor(torch.zeros(x_i), torch.randn(x_v), torch.Size(x_size))
>>> print(x)
torch.sparse.FloatTensor of size (2,2,3) with indices:
tensor([[0]])
and values:
tensor([[[ 0.0832, -0.1464, -0.6159],
         [ 0.0932,  1.3642, -0.2350]]])
>>> x_dense = x.to_dense()
>>> print(x_dense)
tensor([[[ 0.0832, -0.1464, -0.6159],
         [ 0.0932,  1.3642, -0.2350]],

        [[ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000]]])
>>> y = torch.sparse_coo_tensor(torch.zeros(y_i), torch.randn(y_v), torch.Size(y_size))
>>> y_dense = y.to_dense()
>>> x.resize_as_(y)
torch.sparse.FloatTensor of size (2,2) with indices:
tensor([[0]])
and values:
tensor([[ 0.0832, -0.1464]])
>>> x.to_dense()  # The second row doesn't have nonzero values, as expected
tensor([[ 0.0832, -0.1464],
        [ 0.0000,  0.0000]])
>>> x_dense.resize_as_(y_dense)  # Notice that the second row has nonzero values
tensor([[ 0.0832, -0.1464],
        [-0.6159,  0.0932]])
```

Here `x.resize_as_(y)` is actually equivalent to `z=torch.zeros(y_dense.shape); z[0]=x_dense.resize_(y_dense.shape)[0]`, as we need to mask out any changes that are not directly affecting the nnz values.

I think to address the 2nd issue we can perform the resize test on a sparse tensor in which the `values` has the same size as the overall sparse tensor, (e.g. `torch.sparse_coo_tensor(torch.zeros([1, 1]), torch.randn([1, 2, 3]), torch.Size([1, 2, 3]))`), so that we don't need to mask out the zero regions when checking for the correctness of `resize`. Although strictly speaking this is a reduction in coverage and only sufficient because we know how the sparse resize is implemented currently.

For the 1st issue, we could probably set the "data" region to some special value and set everywhere else to 0, so that we can do the equivalence comparison between the two expanded tensors.